lib/wallet/api/http/Cardano/Wallet/Shelley.hs

```diff
@@ -216,13 +217,13 @@ serveWallet
   settings
   tokenMetaUri
   block0
-  beforeMainLoop = evalContT $ do
+  beforeMainLoop = withSNetworkId network $ \sNetwork -> evalContT $  do
```

Comment:
```suggestion
  beforeMainLoop = evalContT $  do
```

I favor *not* bringing `sNetwork` into scope at the beginning of the function — you can do this locally any time. This allows you to keep `netId` in the `where` clause, and avoid introducing additional arguments for the `with*Api` functions. See comments below.

For almost all intents and purposes, `network` and `sNetwork` are the same thing and can be converted into each other at any time.

Reply:
Oh, you're right, sorry. 🙈 Happy to merge as you wrote it.

----

Essentially, the type of `startServer` makes the assertion that all four `ApiLayer` work with the same network (mainnet or a testnet).

One could say that this assertion is now checked on the type-level, at compile-time. Specifically, the function `startServer` can only have this type if it uses the same network id with all four `ApiLayer`. However, we could use the same network id with all four `ApiLayer` *without* putting that fact in the type level — this is what my request for a smaller scope corresponds to. Roughly, this would look something like this:

```hs
startServer ::
  NetworkId
  -> Socket
  -> ApiLayer (RndState) ByronKey 'CredFromKeyK
  -> ApiLayer (SeqState IcarusKey) IcarusKey 'CredFromKeyK
  -> ApiLayer (SeqState ShelleyKey) ShelleyKey 'CredFromKeyK
  -> ApiLayer (SharedState SharedKey) SharedKey 'CredFromScriptK
  -> StakePoolLayer
  -> NtpClient
  -> IO ()
startServer network byron icarus shelley shared … =
    (byron network)
    <|> (icarus network)
    <|> (shelley network)
    <|> (shared network)
```

It's still true that they use the same `network` — it's just that this fact is not checked on the type-level, one has to inspect the definition of `startServer` (value-level) instead. There is no way around inspecting *some code*: Even if we put this property on the type-level, we have to inspect the type of `serveWallet` and check that all four `n` are the same. In this case, I would argue that the amount of human inspection is the same, and we don't get much value out of putting this particular assertion on the type-level. But that is a story for another pull request — these considerations are not in scope here.